src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.BatchingDocumentCollection.cs

```diff
@@ -411,22 +411,18 @@ public async ValueTask ProcessRegularFileChangesAsync(ImmutableArray<string> fil
                         return;
                     }
 
-                    await _project._workspace.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync: true, s =>
+                    await _project._workspace.ApplyBatchChangeToWorkspaceMaybeAsync(useAsync: true, solutionChanges =>
                     {
-                        var accumulator = new SolutionChangeAccumulator(s);
-
                         foreach (var (documentId, textLoader) in documentsToChange)
                         {
-                            if (!s.Workspace.IsDocumentOpen(documentId))
+                            if (!_project._workspace.IsDocumentOpen(documentId))
```

Comment:
here's my terror.  we're in the workspace lock, calling back into the workspace.  How do we know IsDocumentOpen takes no lock itself?

Reply:
So if it takes the main workspace lock a second time, we'll know because it'll deadlock. :smile: I agree though yeah I don't have a good answer for how to enforce the hierarchy, generally though.